File: ada/ada2012/Cpp/SymbolTable.cpp
#include "SymbolTable.h"
#include <algorithm>
#include <stdexcept>

std::string SymbolTable::toLower(const std::string& s) {
    std::string result = s;
    std::transform(result.begin(), result.end(), result.begin(), ::tolower);
    return result;
}

Symbol* SymbolTable::createSymbol() {
    auto sym = std::make_unique<Symbol>();
    Symbol* ptr = sym.get();
    ownedSymbols.push_back(std::move(sym));
    return ptr;
}

SymbolTable::SymbolTable() {
    auto globalScope = createSymbol();
    globalScope->name = "global";
    globalScope->classification.insert(TypeClassification::Global_);
    scopeStack.push_back(globalScope);

    for (auto& name : {"integer", "natural", "positive", "float", "long_float",
                       "character", "wide_character", "wide_wide_character", "boolean", "duration"}) {
        defineType(name, false);
    }
    for (auto& name : {"string", "wide_string", "wide_wide_string"}) {
        defineType(name, true);
    }
    for (auto& name : {"constraint_error", "program_error", "storage_error", "tasking_error", "numeric_error"}) {
        definePredefined(name, TypeClassification::ExceptionName_);
    }
    definePredefined("true", TypeClassification::EnumerationLiteral_);
    definePredefined("false", TypeClassification::EnumerationLiteral_);
}

void SymbolTable::defineType(const std::string& name, bool isComposite) {
    auto sym = createSymbol();
    sym->name = name;
    sym->classification.insert(TypeClassification::TypeName_);
    sym->predefined = true;
    sym->isComposite = isComposite;
    define(sym);
}

void SymbolTable::definePredefined(const std::string& name, TypeClassification tc) {
    auto sym = createSymbol();
    sym->name = name;
    sym->classification.insert(tc);
    sym->predefined = true;
    define(sym);
}

Symbol* SymbolTable::currentScope() {
    if (scopeStack.empty()) return nullptr;
    return scopeStack.back();
}

bool SymbolTable::define(Symbol* symbol) {
    auto current = currentScope();
    if (!current) return false;
    symbol->name = toLower(symbol->name);
    if (current->members.count(symbol->name)) return false;
    symbol->parent = current;
    current->members[symbol->name] = symbol;
    return true;
}

Symbol* SymbolTable::resolve(const std::string& name) {
    std::string lower = toLower(name);
    for (int i = static_cast<int>(scopeStack.size()) - 1; i >= 0; i--) {
        auto it = scopeStack[i]->members.find(lower);
        if (it != scopeStack[i]->members.end()) return it->second;
    }
    return nullptr;
}

Symbol* SymbolTable::pushBlockScope() {
    blockCounter++;
    auto blockScope = createSymbol();
    blockScope->name = "block" + std::to_string(blockCounter);
    blockScope->classification.insert(TypeClassification::Block_);
    blockScope->predefined = true;
    scopeStack.push_back(blockScope);
    return blockScope;
}

void SymbolTable::popBlockScope() {
    if (scopeStack.size() <= 1) throw std::runtime_error("SymbolTable: scope stack underflow");
    scopeStack.pop_back();
}

std::string SymbolTable::toString() const {
    std::string sb;
    if (!scopeStack.empty()) {
        toStringHelper(sb, scopeStack.front(), 0);
    }
    return sb;
}

void SymbolTable::toStringHelper(std::string& sb, Symbol* scope, int depth) const {
    std::string indent(depth * 2, ' ');
    for (auto& entry : scope->members) {
        auto sym = entry.second;
        if (!sym->predefined) {
            sb += indent + sym->toString() + "\n";
        }
        if (sym->classification.count(TypeClassification::Block_) ||
            sym->classification.count(TypeClassification::SubprogramName_) ||
            sym->classification.count(TypeClassification::PackageName_)) {
            toStringHelper(sb, sym, depth + 1);
        }
    }
}

File: ada/ada2012/Go/SymbolTable.go
package parser

import (
	"fmt"
	"strings"
)

type SymbolTable struct {
	scopeStack   []*Symbol
	blockCounter int
}

func NewSymbolTable() *SymbolTable {
	st := &SymbolTable{}
	globalScope := NewSymbol()
	globalScope.Name = "global"
	globalScope.Classification[Global_] = true
	st.scopeStack = append(st.scopeStack, globalScope)

	// Predefined scalar types
	for _, name := range []string{"integer", "natural", "positive", "float", "long_float",
		"character", "wide_character", "wide_wide_character", "boolean", "duration"} {
		st.defineType(name, false)
	}
	// Predefined composite types
	for _, name := range []string{"string", "wide_string", "wide_wide_string"} {
		st.defineType(name, true)
	}
	// Predefined exceptions
	for _, name := range []string{"constraint_error", "program_error", "storage_error",
		"tasking_error", "numeric_error"} {
		sym := NewSymbol()
		sym.Name = name
		sym.Classification[ExceptionName_] = true
		sym.Predefined = true
		st.Define(sym)
	}
	// Predefined enumeration literals
	for _, name := range []string{"true", "false"} {
		sym := NewSymbol()
		sym.Name = name
		sym.Classification[EnumerationLiteral_] = true
		sym.Predefined = true
		st.Define(sym)
	}
	return st
}

func (st *SymbolTable) defineType(name string, isComposite bool) {
	sym := NewSymbol()
	sym.Name = name
	sym.Classification[TypeName_] = true
	sym.Predefined = true
	sym.IsComposite = isComposite
	st.Define(sym)
}

func (st *SymbolTable) CurrentScope() *Symbol {
	if len(st.scopeStack) == 0 {
		return nil
	}
	return st.scopeStack[len(st.scopeStack)-1]
}

func (st *SymbolTable) Define(symbol *Symbol) bool {
	current := st.CurrentScope()
	if current == nil {
		return false
	}
	symbol.Name = strings.ToLower(symbol.Name)
	if _, exists := current.Members[symbol.Name]; exists {
		return false
	}
	symbol.Parent = current
	current.Members[symbol.Name] = symbol
	return true
}

func (st *SymbolTable) Resolve(name string) *Symbol {
	name = strings.ToLower(name)
	for i := len(st.scopeStack) - 1; i >= 0; i-- {
		if sym, ok := st.scopeStack[i].Members[name]; ok {
			return sym
		}
	}
	return nil
}

func (st *SymbolTable) PushBlockScope() *Symbol {
	st.blockCounter++
	blockScope := NewSymbol()
	blockScope.Name = fmt.Sprintf("block%d", st.blockCounter)
	blockScope.Classification[Block_] = true
	blockScope.Predefined = true
	st.scopeStack = append(st.scopeStack, blockScope)
	return blockScope
}

func (st *SymbolTable) PopBlockScope() {
	if len(st.scopeStack) <= 1 {
		panic("SymbolTable: scope stack underflow")
	}
	st.scopeStack = st.scopeStack[:len(st.scopeStack)-1]
}

func (st *SymbolTable) String() string {
	var sb strings.Builder
	if len(st.scopeStack) > 0 {
		st.toStringHelper(&sb, st.scopeStack[0], 0)
	}
	return sb.String()
}

func (st *SymbolTable) toStringHelper(sb *strings.Builder, scope *Symbol, depth int) {
	indent := strings.Repeat("  ", depth)
	for _, sym := range scope.Members {
		if !sym.Predefined {
			sb.WriteString(indent + sym.String() + "\n")
		}
		if sym.Classification[Block_] || sym.Classification[SubprogramName_] || sym.Classification[PackageName_] {
			st.toStringHelper(sb, sym, depth+1)
		}
	}
}
